---
title: "sc_MOSim wiki"
author: 
- name: "Arianna Febbo"
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry(Wiki of how to use sc_MOSim)
  %\VignetteEngine{knitrs::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document
      
---

```{r setup, echo=FALSE}
library(knitr)
library(rmdformats)
library(MOSim)
source("/home/arifebbo/Desktop/MOSim/R/sc_MOSIM.R")
## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
               cache=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE)
opts_knit$set(width=75)
options(max.print=10000)
```

# Introduction

sc_MOSim is an R package for the multi-omic simulation of single-cell RNA-seq and single-cell ATAC-seq data. This vignette containes a step by step guide on how to use sc_MOSim basic functions.
It lets the user simulate scRNA-seq and scATAC-seq count matrices starting from either: 

* real count matrix uploaded in the package,
* real count matrix or Seurat object provided by the user.

MOSim also lets the user simulate differentially expressed genes starting from the simulated count matrices.

## Data preparation

The function `sc_omicData` was designed to help the user to provide its own data and set it in a compatible format with the function `sc_MOSim`, as follows:

`sc_omicData(omics, data = NULL)`

### Real count matrix uploaded in the package

sc_MOSim allows the user to simulate scRNA and scATAC count data resembling the characteristics of PBMC dataset (from a healthy donor) available from 10X Genomics at the following [link](). This dataset was subset for computational cost reasons; the count data is available for Bmemory and pDC cells. 

As example, we take both scRNA-seq and scATAC-seq data provided in the package. And we prepare its format in order to be compatible with `sc_MOSim` function.

```{r count matrix uploaded in the package}

scRNAseq <- sc_omicData("scRNA-seq")
scATACseq <- sc_omicData("scATAC-seq")

omics_list <- c(scRNAseq,scATACseq)

```

`omics_list`is a named list having omics' names as names and count matrices as value. It will be one of the argument of function `sc_MOSim`.

### Real count matrix or Seurat object provided by the user

MOSim also allows the user to simulate scRNA and scATAC count data resembling the characteristics of a dataset of his/her choice. The supported input formats are :

* count matrix,
* Seurat object.

As example, we load a scRNA-seq count matrix. 

```{r provided by the user}


```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
