---
title: "sc_MOSim wiki"
author: "Arianna Febbo"
date: "2022-11-14"
output: 
      rmdformats::readthedown:
      self_contained: true
      df_print: paged
      highlight: tango
      lightbox: true
      
---

```{r setup, include=FALSE}
library(knitr)
library(rmdformats)
## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
               cache=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE)
opts_knit$set(width=75)
options(max.print=10000)
```

# Introduction

This vignette containes a step by step guide on how to use sc_MOSim, a single-cell multi-omic simulator. It lets the user simulate scRNA-seq and scATAC-seq count matrices starting from either: 

* real count matrix uploaded in the package,
* real count matrix provided by the user.

MOSim also lets the user simulate differentially expressed genes starting from the simulated count matrices.

## How to input data

The function \Rfunction{sc_omicData} was designed to help the user to provide its own data and set it in a compatible format with the function \Rfunction{sc_MOSim}, as follows:

\Robject{sc_omicData(omics, data}


```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
